Realm/RLMSyncPermissionValue.mm

```diff
@@ -0,0 +1,203 @@
+        case RLMSyncAccessLevelWrite:
+            return Permission::AccessLevel::Write;
+        case RLMSyncAccessLevelAdmin:
+            return Permission::AccessLevel::Admin;
+    }
+    REALM_UNREACHABLE();
+}
+
+RLMSyncAccessLevel objCAccessLevelForAccessLevel(Permission::AccessLevel level) {
+    switch (level) {
+        case Permission::AccessLevel::None:
+            return RLMSyncAccessLevelNone;
+        case Permission::AccessLevel::Read:
+            return RLMSyncAccessLevelRead;
+        case Permission::AccessLevel::Write:
+            return RLMSyncAccessLevelWrite;
+        case Permission::AccessLevel::Admin:
+            return RLMSyncAccessLevelAdmin;
+    }
+    REALM_UNREACHABLE();
+}
+
+NSString *descriptionForAccessLevel(RLMSyncAccessLevel level) {
+    switch (level) {
+        case RLMSyncAccessLevelNone:
+            return @"none";
+        case RLMSyncAccessLevelRead:
+            return @"read";
+        case RLMSyncAccessLevelWrite:
+            return @"write";
+        case RLMSyncAccessLevelAdmin:
+            return @"admin";
+    }
+    REALM_UNREACHABLE();
+}
+
+// Returns true if the paths are literally equal, or if one path can be translated
+// into the other path via user-ID substitution.
+BOOL pathsAreEquivalent(NSString *thisPath, NSString *thatPath, NSString *thisUserID, NSString *thatUserID)
+{
+    if ([thisPath isEqualToString:thatPath]) {
+        return YES;
+    }
+    NSRange tildeRange = [thisPath rangeOfString:@"~"];
+    if (tildeRange.length > 0) {
+        // Substitute in the user ID for the `/~/` portion of the path, if applicable.
+        return [[thisPath stringByReplacingCharactersInRange:tildeRange
+                                                  withString:thisUserID] isEqualToString:thatPath];
+    }
+    tildeRange = [thatPath rangeOfString:@"~"];
+    if (tildeRange.length > 0) {
+        return [[thatPath stringByReplacingCharactersInRange:tildeRange
+                                                  withString:thatUserID] isEqualToString:thisPath];
+    }
+    return NO;
+}
+
+}
+
+#pragma mark - Permission
+
+@interface RLMSyncPermissionValue () {
+@private
+    NSString *_userID;
+    util::Optional<Permission> _underlying;
+    RLMSyncAccessLevel _accessLevel;
+    NSString *_path;
+}
+@end
+
+@implementation RLMSyncPermissionValue
+
+- (instancetype)initPrivate {
+    self = [super init];
+    return self;
+}
+
+- (instancetype)initWithRealmPath:(NSString *)path
+                           userID:(NSString *)userID
+                      accessLevel:(RLMSyncAccessLevel)accessLevel {
+    if (self = [super init]) {
+        _accessLevel = accessLevel;
+        _path = path;
+        _userID = userID;
+    }
+    return self;
+}
+
+- (instancetype)initWithPermission:(Permission)permission {
+    switch (permission.condition.type) {
+        case ConditionType::UserId:
+            self = [[RLMSyncPermissionValue alloc] initPrivate];
```

Comment:
Yeah, that was cruft left over from when the subclass still existed. Thanks for pointing it out.